### PR TITLE
Use nullptr everywhere possible in library

### PR DESCRIPTION
**Pull request by ZedThree**
Mostly found with clang-tidy, some manual grepping. This replaces `0` and `NULL` with `nullptr` everywhere in the library -- doesn't touch tests or examples -- and uses the compiler so there are no false positives (e.g. use of `NULL` in a string or variable name)

Also apply clang-format to affected lines, along with some manual shifting of comments so it did the sensible thing in some cases.

I'm in two minds about this as it's such an enormous commit, but with a new release coming up, it might be good to get this out of the way.

### CODE REVIEW

**Comment by dschwoerer** [comment]
Having it in several PRs & commits would make it much easier to review ...
I have read the first changes and they seem all sensible, but reviewing everything is challenging.

I am happy for this to go in, but didn't read everything.

File: src/solver/impls/arkode/arkode.cxx [hunk 1]
```diff
@@ -221,17 +221,21 @@ int ArkodeSolver::init(int nout, BoutReal tstep) {
     if(expl){	//Use purely explicit solver
       output.write("\tUsing ARKode Explicit solver \n");
       {TRACE("Calling ARKodeInit");
-	if( ARKodeInit(arkode_mem, arkode_rhs, NULL, simtime, uvec) != ARK_SUCCESS  ) //arkode_rhs_e holds the explicit part, arkode_rhs_i holds the implicit part
-	  throw BoutException("ARKodeInit failed\n");
+        if (ARKodeInit(arkode_mem, arkode_rhs, nullptr, simtime, uvec) !=
```
Comment (bendudson) [inline src/solver/impls/arkode/arkode.cxx]: Not sure this formatting is an improvement; clang seems to get it a bit wrong sometimes

File: src/solver/impls/cvode/cvode.cxx [hunk 2]
```diff
@@ -295,7 +295,7 @@ int CvodeSolver::init(int nout, BoutReal tstep) {
 
     /// Set Jacobian-vector multiplication function
 
-    if((use_jacobian) && (jacfunc != NULL)) {
+    if ((use_jacobian) && (jacfunc != nullptr)) {
```
Comment (bendudson) [inline src/solver/impls/cvode/cvode.cxx]: Above used `jacfunc` rather than `jacfunc != nullptr`

File: src/solver/solver.cxx [hunk 3]
```diff
@@ -529,8 +529,8 @@ int Solver::solve(int NOUT, BoutReal TIMESTEP) {
   
   /// Run the solver
   output_info.write("Running simulation\n\n");
-  
-  time_t start_time = time((time_t*) NULL);
+
+  time_t start_time = time((time_t *)nullptr);
```
Comment (bendudson) [inline src/solver/solver.cxx]: Is the cast needed? Probably doesn't hurt...

File: src/sys/boutexception.cxx [hunk 4]
```diff
@@ -108,7 +108,7 @@ std::string BoutException::BacktraceGenerate() const{
   {                                                                                      \
     buflen = 0;                                                                          \
     buffer = nullptr;                                                                    \
-    if (s == (const char *)NULL) {                                                       \
+    if (s == (const char *)nullptr) {                                                    \
```
Comment (bendudson) [inline src/sys/boutexception.cxx]: Probably don't need cast

File: src/sys/output.cxx [hunk 5]
```diff
@@ -41,7 +41,7 @@ void Output::disable() {
 
 int Output::open(const char *fname, ...) {
 
-  if (fname == (const char *)NULL) {
+  if (fname == (const char *)nullptr) {
```
Comment (bendudson) [inline src/sys/output.cxx]: ditto casts?

**Comment by ZedThree** [comment]
The useless casts will vanish in a future PR :)

**Pull request merged**